File: src/Algorithm/LeetCode/LeetCode_977_Squares_of_a_Sorted_Array/Solution.java
package Algorithm.LeetCode.LeetCode_977_Squares_of_a_Sorted_Array;

class Solution {
    public int[] sortedSquares(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];

        for (int i = 0, j = n - 1, pos = n - 1; i <= j;) {
            if (nums[i] * nums[i] > nums[j] * nums[j]) {
                ans[pos] = nums[i] * nums[i];
                ++i;
            } else {
                ans[pos] = nums[j] * nums[j];
                --j;
            }

            --pos;
        }

        return ans;
    }
}

File: src/Algorithm/LeetCode/LeetCode_977_Squares_of_a_Sorted_Array/Solution.py
from typing import List


# https://www.programmercarl.com/0977.有序数组的平方.html
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        result = [0] * len(nums)
        k = len(nums) - 1

        left = 0
        right = len(nums) - 1

        while left <= right:
            if nums[left] * nums[left] < nums[right] * nums[right]:
                result[k] = nums[right] * nums[right]
                right -= 1
            else:
                result[k] = nums[left] * nums[left]
                left += 1

            k -= 1

        return result
